### ISSUE TITLE
check_swap_1(-1) fails on various architectures

### ISSUE THREAD

**Issue by sebastic** [mapbox/protozero#31]
The Debian package build for protozero 1.2.0 failed on [arm64](https://buildd.debian.org/status/fetch.php?pkg=protozero&arch=arm64&ver=1.2.0-1&stamp=1444632308), [armel](https://buildd.debian.org/status/fetch.php?pkg=protozero&arch=armel&ver=1.2.0-1&stamp=1444632531), [powerpc](https://buildd.debian.org/status/fetch.php?pkg=protozero&arch=powerpc&ver=1.2.0-1&stamp=1444632322), [ppc64el](https://buildd.debian.org/status/fetch.php?pkg=protozero&arch=ppc64el&ver=1.2.0-1&stamp=1444632411), [s390x](https://buildd.debian.org/status/fetch.php?pkg=protozero&arch=s390x&ver=1.2.0-1&stamp=1444632357) & [ppc64](https://buildd.debian.org/status/fetch.php?pkg=protozero&arch=ppc64&ver=1.2.0-1&stamp=1444643593) (both big and little endian architectures) with the same test failure:

```
tests is a Catch v1.2.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
byte swapping
-------------------------------------------------------------------------------
test/t/endian/test_cases.cpp:42
...............................................................................

test/t/endian/test_cases.cpp:45: FAILED:
  REQUIRE( -1 == check_swap_1(-1) )
with expansion:

===============================================================================
test cases:   86 |   85 passed | 1 failed
assertions: 4510 | 4509 passed | 1 failed
```

Please advise how to best help troubleshoot this issue.

**Comment by sebastic** [comment]
The buildlogs for the architectures in question also report:
```
In file included from test/include/test.hpp:2:0,
                 from test/t/endian/test_cases.cpp:6:
test/t/endian/test_cases.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____42()':
test/t/endian/test_cases.cpp:45:34: warning: negative integer implicitly converted to unsigned type [-Wsign-conversion]
     REQUIRE(-1 == check_swap_1(-1));
                                  ^
test/t/endian/test_cases.cpp:45:16: warning: comparison is always false due to limited range of data type [-Wtype-limits]
     REQUIRE(-1 == check_swap_1(-1));
                ^
```

**Comment by daniel-j-h** [comment]
In https://github.com/mapbox/protozero/blob/1bca7abc2c12ca985b290d682d32821f47ff775b/test/t/endian/test_cases.cpp#L10: it is implementation defined if char is signed or unsigned. Integral literals are signed.

**Comment by sebastic** [comment]
Change the type to `uint8_t` or `int8_t` (which seems more appropriate for `-1`) causes a 'no matching function for call' error:
```
test/t/endian/test_cases.cpp: In function 'char {anonymous}::check_swap_1(int8_t)':
test/t/endian/test_cases.cpp:14:47: error: no matching function for call to 'byteswap(int8_t*, char*)'
         protozero::byteswap<1>(&data, &swapped);
                                               ^
In file included from test/t/endian/test_cases.cpp:4:0:
include/protozero/byteswap.hpp:29:13: note: candidate: template<int N> void protozero::byteswap(const char*, char*)
 inline void byteswap(const char* /*data*/, char* /*result*/) {
             ^
include/protozero/byteswap.hpp:29:13: note:   template argument deduction/substitution failed:
test/t/endian/test_cases.cpp:14:47: note:   cannot convert '& data' (type 'int8_t* {aka signed char*}') to type 'const char*'
         protozero::byteswap<1>(&data, &swapped);
                                               ^
In file included from test/include/test.hpp:2:0,
                 from test/t/endian/test_cases.cpp:6:
test/t/endian/test_cases.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____42()':
test/t/endian/test_cases.cpp:45:16: warning: comparison is always false due to limited range of data type [-Wtype-limits]
     REQUIRE(-1 == check_swap_1(-1));
                ^
```

**Comment by daniel-j-h** [comment]
@sebastic sorry for the confusion --- you can not simply change the type just on the call site. Just wanted to add an idea how to solve this, as I was reading this issue by accident.

**Comment by sebastic** [comment]
FWIW, this is how Boost 1.58 does a uint8_t swap (in [boost/endian/conversion.hpp](http://sources.debian.net/src/boost1.58/1.58.0%2Bdfsg-3/boost/endian/conversion.hpp/?hl=239#L239)):
```c++
  inline uint8_t endian_reverse(uint8_t x) BOOST_NOEXCEPT
  {
    return x;
  }
```
This is quite similar to `byteswap<1>` in protozero, but Boost never uses chars in its signatures.

**Comment by joto** [comment]
This code triggering this is never actually used, so I removed it and the test with it. Thanks for catching this. I released a new version 1.2.1.
